### PR TITLE
denc: fix internal fragmentation when decoding ptr in bl

### PR DESCRIPTION
**Pull request by tchaikov**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by ukernel** [comment]
```
mds: label old_inode_t and frag_t with denc_traits

types which do not provide denc_traits use the default denc_traits,
which has need_contiguous=false. 
```

I think need_contiguous is true by default

**Comment by ukernel** [comment]
Tested by me, it resolved memory fragment issue when decoding log event

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-11-27_11:44:27-rados-wip-kefu2-testing-2018-11-27-1724-distro-basic-smithi/

File: src/include/denc.h [hunk 1]
```diff
@@ -713,17 +713,28 @@ struct denc_traits<std::basic_string<char,std::char_traits<char>,A>> {
...
+  static void decode_nohead(size_t len, value_type& s,
+                            buffer::list::const_iterator& p) {
+    s.clear();
+    if (len) {
+      if constexpr (std::is_same_v<value_type, std::string>) {
+        s.clear();
+        p.copy(len, s);
+      } else {
+        s.append(len, 0);
+        p.copy(len, s.data());
```
Comment (cbodley) [inline src/include/denc.h]: the std::string case duplicates the call to clear, and this case could replace clear+append with resize

Comment (tchaikov) [reply src/include/denc.h]: @cbodley thanks! fixed and repushed.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/sage-2018-11-28_02:28:18-rados-wip-sage2-testing-2018-11-27-1650-distro-basic-smithi/

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-11-29_01:22:06-rados-wip-kefu2-testing-2018-11-29-0053-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-11-29_08:14:41-rados-wip-kefu2-testing-2018-11-29-1356-distro-basic-smithi/

test failures are tracked / caused by

- #19454
- http://tracker.ceph.com/issues/37391

**Pull request merged**